From 37c79f8e2975110a40cc6679877019262c3d2b4e Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Fri, 25 Aug 2017 14:37:20 +0200
Subject: [PATCH 01/15] mmc/sdhci: Add card detect method

Detect whether a card is detected. This prevents probing problems
when no card is in the reader. The board apparently reports "present"
for the non-removable device as well, so we can get away with that.

With "non-removable" in the devicetree there is no card detect. Pass
this information on as a quirk (SDHCI_QUIRK_NO_CD) and use it in the
sdhci driver.
---
 drivers/mmc/sdhci.c      | 13 +++++++++++++
 drivers/mmc/zynq_sdhci.c |  3 +++
 include/sdhci.h          |  1 +
 3 files changed, 17 insertions(+)

diff --git a/drivers/mmc/sdhci.c b/drivers/mmc/sdhci.c
index 628bd7a6da..6d2caf0ab5 100644
--- a/drivers/mmc/sdhci.c
+++ b/drivers/mmc/sdhci.c
@@ -565,6 +565,17 @@ static int sdhci_init(struct mmc *mmc)
 	return 0;
 }
 
+static int sdhci_get_cd(struct udevice *dev)
+{
+	struct mmc *mmc = mmc_get_mmc_dev(dev);
+	struct sdhci_host *host = mmc->priv;
+
+	if (host->quirks & SDHCI_QUIRK_NO_CD)
+		return -ENOSYS;
+
+	return !!(sdhci_readl(host, SDHCI_PRESENT_STATE) & SDHCI_CARD_PRESENT);
+}
+
 #ifdef CONFIG_DM_MMC
 int sdhci_probe(struct udevice *dev)
 {
@@ -610,12 +621,14 @@ const struct dm_mmc_ops sdhci_ops = {
 #ifdef MMC_SUPPORTS_TUNING
 	.execute_tuning	= sdhci_execute_tuning,
 #endif
+	.get_cd         = sdhci_get_cd,
 };
 #else
 static const struct mmc_ops sdhci_ops = {
 	.send_cmd	= sdhci_send_command,
 	.set_ios	= sdhci_set_ios,
 	.init		= sdhci_init,
+	.get_cd         = sdhci_get_cd,
 };
 #endif
 
diff --git a/drivers/mmc/zynq_sdhci.c b/drivers/mmc/zynq_sdhci.c
index 2293964b82..ca7bd2b647 100644
--- a/drivers/mmc/zynq_sdhci.c
+++ b/drivers/mmc/zynq_sdhci.c
@@ -408,6 +408,9 @@ static int arasan_sdhci_probe(struct udevice *dev)
 	host->quirks = SDHCI_QUIRK_WAIT_SEND_CMD |
 		       SDHCI_QUIRK_BROKEN_R1B;
 
+	if (fdtdec_get_bool(gd->fdt_blob, dev_of_offset(dev), "non-removable"))
+		host->quirks |= SDHCI_QUIRK_NO_CD;
+
 #ifdef CONFIG_ZYNQ_HISPD_BROKEN
 	host->quirks |= SDHCI_QUIRK_BROKEN_HISPD_MODE;
 #endif
diff --git a/include/sdhci.h b/include/sdhci.h
index cc2a1fe76b..7581418ca0 100644
--- a/include/sdhci.h
+++ b/include/sdhci.h
@@ -244,6 +244,7 @@
 #define SDHCI_QUIRK_WAIT_SEND_CMD	(1 << 6)
 #define SDHCI_QUIRK_USE_WIDE8		(1 << 8)
 #define SDHCI_QUIRK_NO_1_8_V		(1 << 9)
+#define SDHCI_QUIRK_NO_CD		(1 << 11)
 
 /*
  * mmc host capabilities
-- 
2.17.1

